gcn_monitor/kafka.py

```diff
@@ -23,7 +26,70 @@ def stats_cb(data):
         metrics.broker_state.labels(broker["name"]).state(broker["state"])
 
 
-def run():
+def sanitize_topic(topic):
+    """Replaces invalid characters in the topic string so it can be used as a file name
+
+    Args:
+        topic (string): Kafka message topic
+
+    Returns:
+        safe_name: A formatted version of the topic string with underscores(`_`) in place of illegal characters
+    """
+    invalid_chars = r'[\/:*?"<>|]'
+    safe_name = re.sub(invalid_chars, "_", topic)
+    return safe_name.strip()
```

Comment:
This is a one-way transformation. We can't invert it uniquely. Is this really the default behavior for the connector that we are basing this on? Do we have enough information in the data files to reconstruct the actual topic name if we ever need to restore it?

Reply:
I actually don't remember seeing anything in the connector that did this. I think the idea was just to replace characters that will mess with the formatting of the string as a URL

Reply:
The project is here: https://github.com/confluentinc/kafka-connect-storage-cloud

Would you please check how it encodes filenames?

Reply:
It looks like [this](https://github.com/confluentinc/kafka-connect-storage-common/blob/master/partitioner/src/main/java/io/confluent/connect/storage/partitioner/DefaultPartitioner.java) is the definition for the partition rules. Between the code in the link ^ and this one, I am not seeing anything that is looking like they transform the topic

Reply:
In that case, the Kafka connector may be vulnerable to [CWE-641: Improper Restriction of Names for Files and Other Resources](https://cwe.mitre.org/data/definitions/641.html). We should talk to Confluent about that.

Reply:
@lpsinger actually I think we may be able to just rely on the valid legal characters for kafka topics
https://github.com/apache/kafka/blob/0.10.2/core/src/main/scala/kafka/common/Topic.scala#L29C13-L29C13

Reply:
Ah, that's great. Where are those restrictions documented?

Reply:
I'm not finding literal documentation on it, but mainly just references to their source code from blog posts. One specifically went here


https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/internals/Topic.java#L75

```java
    private static String detectInvalidTopic(String name) {
        if (name.isEmpty())
            return "the empty string is not allowed";
        if (".".equals(name))
            return "'.' is not allowed";
        if ("..".equals(name))
            return "'..' is not allowed";
        if (name.length() > MAX_NAME_LENGTH)
            return "the length of '" + name + "' is longer than the max allowed length " + MAX_NAME_LENGTH;
        if (!containsValidPattern(name))
            return "'" + name + "' contains one or more characters other than " +
                "ASCII alphanumerics, '.', '_' and '-'";
        return null;
    }
```

Reply:
I agree. So do you need to transform the topic name at all?

Please add a comment to the code to explain that Kafka constrains the topic name to contain only characters that are safe to include in filenames.

Would you please submit an issue for Apache Kafka to document the topic name constraints?

Reply:
I think we don't need the transform, I'll remove that and the comment in the next commit.

And yup for the issue